src/track/serato/markers.cpp

```diff
@@ -0,0 +1,179 @@
+#include "track/serato/markers.h"
+
+#include <QtEndian>
+#include <QStringLiteral>
+
+namespace {
+
+const int kEntrySize = 22;
+const char* kVersion = "\x02\x05";
+const int kVersionSize = 2;
+
+inline
+quint32 bytesToUInt32(const QByteArray& data) {
+    DEBUG_ASSERT(data.size() == sizeof(quint32));
+#if QT_VERSION >= QT_VERSION_CHECK(5, 12, 0)
+    return qFromBigEndian<quint32>(data.constData());
+#else
+    return qFromBigEndian<quint32>(
+            reinterpret_cast<const uchar*>(data.constData()));
+#endif
+}
+
+QRgb colorToRgb(quint8 w, quint8 x, quint8 y, quint8 z) {
+    quint8 b = (z & 0x7F) | ((y & 0x01) << 7);
+    quint8 g = ((y & 0x7F) >> 1) | ((x & 0x03) << 6);
+    quint8 r = ((x & 0x7F) >> 2) | ((w & 0x07) << 5);
+    return QRgb(r << 16) | (g << 8) | b;
+}
+
+QRgb colorToRgb(const QByteArray& data) {
+    DEBUG_ASSERT(data.size() == 4);
+    return colorToRgb(data.at(0), data.at(1), data.at(2), data.at(3));
+}
+
+quint32 colorFromRgb(quint8 r, quint8 g, quint8 b) {
+    quint8 z = b & 0x7F;
+    quint8 y = ((b >> 7) | (g << 1)) & 0x7F;
+    quint8 x = ((g >> 6) | (r << 2)) & 0x7F;
+    quint8 w = (r >> 5);
+    return (w << 24) | (x << 16) | (y << 8) | z;
+}
+
+quint32 colorFromRgb(QRgb rgb) {
+    return colorFromRgb(qRed(rgb), qGreen(rgb), qBlue(rgb));
+}
+
+}
+
+namespace mixxx {
+
+QByteArray SeratoMarkersEntry::data() const {
+    QByteArray data;
+    data.resize(kEntrySize);
+
+    QDataStream stream(&data, QIODevice::WriteOnly);
+    stream.setVersion(QDataStream::Qt_5_0);
+    stream.setByteOrder(QDataStream::BigEndian);
+    stream << (quint8)(m_isSet ? '\x00' : '\x7F')
+           << (quint32)((m_startPosition == -1) ? 0x7F7F7F7F : m_startPosition)
+           << (quint8)((!m_isSet || m_type == 1) ? '\x7F' : m_isEnabled)
+           << (quint32)((m_endPosition == -1) ? 0x7F7F7F7F : m_endPosition);
+    stream.writeRawData("\x00\x7F\x7F\x7F\x7F\x7F", 6);
+    stream << (quint32)colorFromRgb(m_color)
+           << (quint8)m_type
+           << (quint8)m_isLocked;
+    return data;
+}
+
+SeratoMarkersEntry* SeratoMarkersEntry::parse(const QByteArray& data) {
+    if (data.length() != kEntrySize) {
+        qWarning() << "Parsing SeratoMarkersEntry failed:"
+                   << "Length" << data.length() << "!=" << kEntrySize;
+        return nullptr;
+    }
+
+    const bool isSet = (data.at(0) == '\x7F') ? false : true;
+    const quint32 uStartPosition = bytesToUInt32(data.mid(1, 4));
```

Comment:
Are you sure? At least for QByteArray's copy constructor the documentation says:
> This operation takes constant time, because QByteArray is implicitly shared. This makes returning a QByteArray from a function very fast. If a shared instance is modified, it will be copied (copy-on-write), taking linear time

Anyway, I replaced QByteArray with QDataStream in `SeratoMarkersEntry::parse`.

Reply:
I also replaced it in `SeratoMarkers::parse` now. We still do a deep copy of each entry, but I can't change that without modifying the function signature :-/

Reply:
> Are you sure?

Yes, because mid() and friends can't implicit share, because they create a new, smaller Array.